File: 94.二叉树的中序遍历.java
import java.sql.Array;
import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

import javax.swing.tree.TreeNode;

/*
 * @lc app=leetcode.cn id=94 lang=java
 *
 * [94] 二叉树的中序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /*
     * 递归 时间O(n) 空间 O(1)
    */
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        accessTree(root, res);
        return res;
    }

    public void accessTree(TreeNode root,List<Integer>res){
        if (root==null){
            return ;
        }
        accessTree(root.left,res);
        res.add(root.val);
        accessTree(root.right,res);
    }
    

    /*
     * 循环迭代
     */
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer>res =new ArrayList<>();
        Deque <TreeNode> stack = new LinkedList<>();
        while (root!=null||!stack.isEmpty()){
            while (root!=null){
            stack.push(root);
            root=root.left;
            }
            root=stack.pop();
            res.add(root.val);
            root=root.right;
        }
        return res;
    }
}
// @lc code=end

File: 94.二叉树的中序遍历.js
/*
 * @lc app=leetcode.cn id=94 lang=javascript
 *
 * [94] 二叉树的中序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
// var inorderTraversal = function (root) {
//   if (root == null) return [];
//   let res = new Array();
//   const dfs = (root) => {
//     if (root == null) return;
//     dfs(root.left);
//     res.push(root.val);
//     dfs(root.right);
//   };
//   dfs(root);
//   return res;
// };

var inorderTraversal = function (root) {
  // 入栈： 左  右
  //出： 中 左 右
  if (root == null) return [];
  let stack = [];
  let res = [];
  let cur = root;
  while (cur || stack.length) {
    if (cur) {
      stack.push(cur);
      cur = cur.left;
    } else {
      cur = stack.pop();
      res.push(cur.val);
      cur = cur.right;
    }
  }
  return res;
};
// @lc code=end
